### PR TITLE
LibGit2 credential approval/rejection

### PR DESCRIPTION
**Pull request by omus**
Previously the credential callback would mutate the explicitly provided credential or cached credential. With this change we never mutate explicitly provided credentials and only add/replace credentials in the cache when they have successfully been used.

Additionally, this PR lays the ground work for integrating other credential storage systems such as the [git credential helpers](https://git-scm.com/docs/gitcredentials#gitcredentials-helper) and removes the LibGit2 function `get_creds!`.

Part of #20725.

### CODE REVIEW

**Comment by omus** [comment]
I missed a usage of `get_creds!` in test/misc.jl

File: base/libgit2/callbacks.jl [hunk 1]
```diff
@@ -45,7 +45,7 @@ function user_abort()
           (Cint, Cstring),
           Cint(Error.Callback), "Aborting, user cancelled credential request.")
 
-    return Cint(Error.EAUTH)
+    return Cint(Error.EUSER)
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: 👍

File: base/libgit2/types.jl [hunk 2]
```diff
@@ -1119,14 +1119,29 @@ struct CachedCredentials <: AbstractCredentials
     CachedCredentials() = new(Dict{String,AbstractCredentials}())
 end
 
-"Obtain the cached credentials for the given host+protocol (credid), or return and store the default if not found"
-get_creds!(collection::CachedCredentials, credid, default) = get!(collection.cred, credid, default)
```
Comment (rofinn) [inline base/libgit2/types.jl]: Should probably deprecate `get_creds!` properly.

**Review by rofinn — CHANGES_REQUESTED** [review]
Apart from properly deprecating `get_creds!` this seems reasonable.

File: base/libgit2/types.jl (hunk 2 above)
Comment (omus) [reply base/libgit2/types.jl]: :+1:

Comment (omus) [reply base/libgit2/types.jl]: Added a proper deprecation for this. It required adding [`Base.get!(cache::CachedCredentials, cred_id, default)`](https://github.com/JuliaLang/julia/pull/23711/files#diff-13a2010bd5b11c9f5d12b75cfba76bc9R1124). Additionally, I switched the test/misc.jl tests to use `get!`.

File: base/libgit2/callbacks.jl (hunk 1 above)
Comment (omus) [reply base/libgit2/callbacks.jl]: Just going to mention that [`EUSER`](https://github.com/JuliaLang/julia/blob/master/base/libgit2/error.jl#L13) is a generic error code which is suppose to be for callback usage. It doesn't formally mean the user cancelled the callback.

**Pull request merged**